### ISSUE TITLE
400 Bad request on PSU Control Calls from WebIF

### ISSUE THREAD

**Issue by itsyoshio** [OctoFarm/OctoFarm#777]
### OctoFarm Version

1.1.13-hotfix

### OctoPrint Version(s)

1.7.2

### pm2 or docker?

pm2

### Operating System

FarmPi (Ubuntu 20.04.2 LTS)

### OctoPrint Plugins

PSU Control

### Location of the problem

Web Interface

### Description of the problem

```bash
PSU Control Api requests get 400 Bad Requests

OctoFarm is on a Raspberry Pi 3 (installed via FarmPi)\
on that same Raspberry Pi 3, there is a (manual) install of OctoPrint.

OctoPrint has the PSU Plugin, which works in the OctoPrint Webinterface.\
The PSU Plugin also works with curl (Rest API on OctoPrint works).

Requests to control the PSU with the OctoFarm UI result in 400 Bad Requests regardless of how i configure it (i did Empty Cache and Hard Reload after changing the Power Settings).

CORS is Enabled, Controlling Works, Printing Works, Upload Works.
```

### Steps taken to reproduce

```bash
Set up Power Control in OctoFarm

Versions i've tried : 

`[PrinterURL]/api/plugin/psucontrol?apikey=[PrinterAPI]`

`http://localhost:80/octoprint/api/plugin/psucontrol?apikey=[PrinterAPI]`
(Afterwards i've realised the request is a JS Fetch request made from my Client, not the Server)

`http://192.168.188.50/octoprint/api/plugin/psucontrol?apikey=[PrinterAPI]`

each with and without APIkey.
```

### What did you expect to happen

```bash
Turning PSU On, Off, Toggling and getting State in OctoFarm.
```

### To avoid the problem I had to

```bash
No resolution yet
```
[OctoFarm-Server.log](https://github.com/OctoFarm/OctoFarm/files/7676592/OctoFarm-Server.log)

[JS Console Log.txt](https://github.com/OctoFarm/OctoFarm/files/7676611/JS.Console.Log.txt)

<details>
  <summary>JS Console</summary>

  ```javascript
    printerManagementRunner.min.js:8434 POST http://192.168.188.50/octoprint/api/plugin/psucontrol?apikey=<KEY> 400 (BAD REQUEST)
    (anonymous) @ printerManagementRunner.min.js:8434
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:7700
    s @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:8473
    (anonymous) @ printerManagementRunner.min.js:3199
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3222
    printerManagementRunner.min.js:8535 POST http://192.168.188.50/octoprint/api/plugin/psucontrol?apikey=<KEY> 400 (BAD REQUEST)
    (anonymous) @ printerManagementRunner.min.js:8535
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:7700
    s @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:8589
    (anonymous) @ printerManagementRunner.min.js:3203
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    Promise.then (async)
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3222
    printerManagementRunner.min.js:8535 POST http://192.168.188.50/octoprint/api/plugin/psucontrol?apikey=<KEY> 400 (BAD REQUEST)
    (anonymous) @ printerManagementRunner.min.js:8535
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:7700
    s @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:8589
    (anonymous) @ printerManagementRunner.min.js:3207
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    Promise.then (async)
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    Promise.then (async)
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3222
    printerManagementRunner.min.js:8535 POST http://192.168.188.50/octoprint/api/plugin/psucontrol?apikey=<KEY> 400 (BAD REQUEST)
    (anonymous) @ printerManagementRunner.min.js:8535
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:7700
    s @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:8589
    (anonymous) @ printerManagementRunner.min.js:3211
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    Promise.then (async)
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    Promise.then (async)
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    Promise.then (async)
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3222
    printerManagementRunner.min.js:8535 POST http://192.168.188.50/octoprint/api/plugin/psucontrol?apikey=<KEY> 400 (BAD REQUEST)
    (anonymous) @ printerManagementRunner.min.js:8535
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:7700
    s @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:8589
    (anonymous) @ printerManagementRunner.min.js:3238
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3368
    (anonymous) @ printerManagementRunner.min.js:3420
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3451
    (anonymous) @ printerManagementRunner.min.js:8982
    C.onmessage @ printerManagementRunner.min.js:8981
    serverAliveCheck.min.js:609 Fetch finished loading: GET "http://192.168.188.50/serverChecks/amialive".
    (anonymous) @ serverAliveCheck.min.js:609
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ serverAliveCheck.min.js:472
    c @ serverAliveCheck.min.js:474
    (anonymous) @ serverAliveCheck.min.js:474
    (anonymous) @ serverAliveCheck.min.js:474
    printerManagementRunner.min.js:8535 POST http://192.168.188.50/octoprint/api/plugin/psucontrol?apikey=<KEY> 400 (BAD REQUEST)
    (anonymous) @ printerManagementRunner.min.js:8535
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:7700
    s @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:7702
    (anonymous) @ printerManagementRunner.min.js:8589
    (anonymous) @ printerManagementRunner.min.js:3238
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3368
    (anonymous) @ printerManagementRunner.min.js:3420
    u @ printerManagementRunner.min.js:18971
    (anonymous) @ printerManagementRunner.min.js:19201
    (anonymous) @ printerManagementRunner.min.js:19026
    i @ printerManagementRunner.min.js:3057
    s @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3059
    (anonymous) @ printerManagementRunner.min.js:3451
    (anonymous) @ printerManagementRunner.min.js:8982
    C.onmessage @ printerManagementRunner.min.js:8981
```
  
</details>

**Comment by NotExpectedYet** [comment]
I have documentation for those here. 

https://docs.octofarm.net/guides/octoprint-supported-plugins.html#octoprint-s-power-control-plugins

PSU control you can use the examples on the power settings page. You need to paste in those commands though it won't work without those hence the 400 error. Hope that helps.

**Comment by NotExpectedYet** [comment]
Just to add I don't think you should have to nuke cache. The commands are built dynamically so should work. The only limitation I remember was you'd need a page refresh to make the button show up.

**Comment by NotExpectedYet** [comment]
Out of curiosity how's the b3+ been handling OP and OF? 

If you've any thoughts on making that screen clearer then let me know :). Next version the auto pulling of those settings is fixed so anyone with PSU control already setup when OF scans won't hit this.

**Comment by itsyoshio** [comment]
### Fetch Requests
Oh, i would've caught that if i had actually tested POST requests with curl, so sorry about that!\
I i didn't even think about the possibility that i'm using incorrect JSON in my POST.\
Not to blame you at all, but the help text below is a bit misleading, maybe change the help text to a link for the [Docs](https://docs.octofarm.net/guides/octoprint-supported-plugins.html#octoprint-s-power-control-plugins)?

### Docs
The Docs are really helpfull. Sorry i didn't think of searching the Website, my Google Skills didn't yield any results.

### Ticket
I'll go ahead and close this, i'll only be able to implement this on Wednesday since this is running in my University MakerSpace and that is my next Shift. I'll reopen in the most unlikely event that didn't resolve my issue.

### Usage
Regarding the Handling : i have not looked at the actual cpu utilization or mem yet, but i will do so and report back.\
Printing with a Job from OctoFarm did result in regular, stable printing.

Actually we have 4 Pi's in a Eletric Box with Dual SSR Relais for each Channel / Printer.\
In the End one OctoFarm will oversee 4 OctoPrints (where OctoFarm is on the same Pi as one OctoPrint).

Once that's done (probably next Wednesday) i will report back with more Stats.

---

Thank you so much for your help.\
Have a great Day.

**Comment by NotExpectedYet** [comment]
> ### Fetch Requests
> 
> Oh, i would've caught that if i had actually tested POST requests with curl, so sorry about that! I i didn't even think about the possibility that i'm using incorrect JSON in my POST. Not to blame you at all, but the help text below is a bit misleading, maybe change the help text to a link for the [Docs](https://docs.octofarm.net/guides/octoprint-supported-plugins.html#octoprint-s-power-control-plugins)?

Tbh, A little button with those images that pop up might be nice, one for each supported plugin. This is exactly why I asked, those interface notes haven't been touched since I first wrote them so I thought they might be a bit misleading. If the automation grabbing of settings was working we wouldn't be here regardless! aha. 

No trouble on the rest either, do let me know if it resolves with those objects. 

That's good to know regardless, I've had some issues with increased CPU over a Pi which is resolved in V1.2 but if you've not hit it so far doubt you will.
